### DISCUSSION TITLE
How to make a complex update happen in-place

### DISCUSSION THREAD

**Discussion by hr0nix** [jax-ml/jax#12209] [Q&A]
Consider the following code (which is an illustrative example inspired by an actual problem):
```python
def make_item(n):
    return jnp.full((64,), n)

def make_item_batch(n, batch_size):
    return jnp.full((2, batch_size, 64,), n)

def modify_slice(slice):
    return slice.at[10].set(make_item(1))

def select_index(state):
    return 0
    
@partial(jax.jit, donate_argnums=(0,))
def update_func(state):
    index = select_index(state)
    state_slice = state[index]
    state_slice = modify_slice(state_slice)
    state = state.at[index].set(state_slice)
    return state

state = make_item_batch(0, 1_000_000)
state = update_func(state)  # I would like this update to happen in-place
```
I would like the call to `update_func` to update `state[index][10]` in-place, however this is not what's currently happening in JAX (can be checked by benchmarking `update_func`). What I presume is happening is that whole `state[index]` gets updated in-place, but not `state[index][10]`.

The update does happen in-place, however, If I rewrite `update_func` like this:
```python
@partial(jax.jit, donate_argnums=(0,))
def update_func(state):
    index = select_index(state)
    state = state.at[index, 10].set(make_item(1))
    return state
```

My problem is that in practice I cannot rewrite my `update_func` as above, because I need it to be implemented in terms of `modify_slice`, which requires a slice of the array as an input.

So two questions here:
* Is it expected that JAX doesn't recognize a possibility for a smaller in-place update in the first case?
* Is there any way to rewrite this code such that only `state[index][10]` gets updated, but the code uses `modify_slice` to do the update.

**EDIT:** clarified the question in response to @jakevdp comment.

**Comment by jakevdp** [comment]
Can you say more about how you determined the update was not happening inplace? If I run either version of your code on a Colab GPU runtime (note that buffer donation is [not supported](https://jax.readthedocs.io/en/latest/faq.html#buffer-donation) on CPU) I get the following:
```python
state = make_item_batch(0, 1_000_000)
buffer = state.unsafe_buffer_pointer()
state = update_func(state)
print(buffer == state.unsafe_buffer_pointer())
# True
```
It looks to me like the update is happening in-place, unless there's an aspect of this that I'm missing.

**Reply by hr0nix** [reply]
I’ve measured the time it takes to execute the code in both cases. I think what might be happening is that update is indeed in-place, but the whole `state[index]` gets overwritten instead of just `state[index, 10]`, which is the desired behavior.

**Reply by hr0nix** [reply]
@jakevdp I hopefully clarified the question a bit. Any suggestions on how to achieve the desired behavior?

**Reply by jakevdp** [reply]
Hi - thanks for the update. I believe this kind of decision (whether to do things inplace or with temporary memory) is made within the XLA compiler, and I don't believe there is any Python API to control how the compiler makes tradeoffs between computation cost and memory use.

**Reply by hr0nix** [reply]
I'd like to point out that there is no tradeoff here, it is more of a missed optimization opportunity.

Can you advise on how can I raise this issue with the XLA compiler team? Will they accept JAX code as a repro case?
